vnext/JSI/Shared/ChakraObjectRef.h

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "jsi/jsi.h"
+
+#ifdef CHAKRACORE
+#include "ChakraCore.h"
+#else
+#ifndef USE_EDGEMODE_JSRT
+#define USE_EDGEMODE_JSRT
+#endif
+#include <jsrt.h>
+#endif
+
+#include <exception>
+#include <string>
+
+namespace Microsoft::JSI {
+
+inline void CrashUponChakraError(JsErrorCode error) {
+  if (error != JsNoError) {
+    std::terminate();
+  }
+}
+
+void ThrowUponChakraError(
+    JsErrorCode error,
+    const char *const chakraApiName = nullptr);
+
+// An shared_ptr like RAII Wrapper for JsRefs, which are references to objects
+// owned by the garbage collector. These include JsContextRef, JsValueRef, and
+// JsPropertyIdRef, etc. ChakraObjectRef ensures that JsAddRef and JsRelease
+// are called upon initialization and invalidation, respectively. It also allows
+// users to implicitly convert it into a JsRef. A ChakraObjectRef must only be
+// initialized once and invalidated once.
+class ChakraObjectRef {
+ public:
+  ChakraObjectRef() noexcept {}
+  inline explicit ChakraObjectRef(JsRef ref) {
+    Initialize(ref);
+  }
+
+  ChakraObjectRef(const ChakraObjectRef &original) noexcept;
+  ChakraObjectRef(ChakraObjectRef &&original) noexcept;
+
+  ChakraObjectRef &operator=(const ChakraObjectRef &rhs) noexcept;
+  ChakraObjectRef &operator=(ChakraObjectRef &&rhs) noexcept;
+
+  ~ChakraObjectRef() noexcept;
+
+  void Initialize(JsRef ref);
+  void Invalidate();
+
+  inline operator JsRef() const noexcept {
+    return m_ref;
+  }
+
+ private:
+  void swap(ChakraObjectRef &other);
```

Comment:
>swap [](start = 7, length = 4)

(nit) why case change?

Reply:
That's a mistake on my part. It will be fixed in a future iteration.

---
In reply to: [336144480](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336144480) [](ancestors = 336144480)